trivial-builders: replace writeReferencesToFile with writeClosure

ShamrockLee opened the pull request:
###### Description of changes

The new builder takes a list `paths` instead of a single path, and print their closure to a file. It would be helpful when building a container image from a list of
packages. See https://github.com/NixOS/nixpkgs/issues/177908#issuecomment-1160624616 .

I have prepared two versions of implementation, both of which tries to merge the result of `writeReferencesToFile` on each derivation:

1.  Merge into a Bash associative array with the out-path hashes as the keys, and sort the keys lexicographically with the `sort` command. Time complexity: $O(N \log N)$, I/O time complexity $O(N)$, where $N$ is the total number of references. (https://github.com/ShamrockLee/nixpkgs/blob/write-multiple-references-dict/pkgs/build-support/trivial-builders.nix#L671=)
2.  Merge using the merge-sorted-arrays algorithm given that the results of `writeReferencesToFile` are sorted. Time complexity: $O(mN)$, I/O time complexity: $O(N)$, where $m$ is the length of `paths` and $N$ is the total number of references. (https://github.com/ShamrockLee/nixpkgs/blob/write-multiple-references-merge/pkgs/build-support/trivial-builders.nix#L671=)

This PR adopts the first implementation, since the code is shorter and more understandable.

Update:
@SomeoneSerge proposed the third approach: Write the path list with `writeText`, get its list of dependencies with `writeReferencesToFile` and then remove the line containing the store path of the path list derivation. Just re-implement this PR following this approach. 

Update:
Use the new `exportReferencesGraph` interface with `__structuredAttrs = true;` and parse `.attrs.json` with `jq`.

~~The test cases of `writeReference*` builders doesn't run (#204456), blocking the test case adding for this new builder.~~

Update:

The test that makes sure the equivalence between the `writeMultipleReferencesToFile [ path1 path2 ... ]` and `nix-store --query --references outPath1 outPath2 ...` is added to the NixOS test `tests.trivial-builders.references`.

Add two package tests:
* `tests.trivial-builders.writeMultipleReferencesToFile-single` ensure the equivalence between `writeMultipleReferencesToFile [ path ]` and `writeReferencesToFile path`.
* `tests.trivial-builders.writeMultipleReferencesToFile-mixed` ensures that the content of the result derivation of `writeMultipleReferencesToFile [ path1 path2 ... ]` is the unique collection of that of (`writeReferencesToFile path1`, `writeReferencesToFile path2`, ...).

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
  - Content of files produced by by `pkgs.writeMultipleReferencesToFile [ pkgs.jupyter ]` and `pkgs.writeReferencesToFile pkgs.jupyter` is the same.
  - Content of files produced by by `pkgs.writeMultipleReferencesToFile [ pkgs.gcc pkgs.jupyter ]` and `pkgs.writeReferencesToFile [ pkgs.jupyter pkgs.gcc ]` is the same.
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ShamrockLee commented:
@jbedo @SomeoneSerge @GaetanLepage

This could replace `singularity-tools.mkLayer` and prevent unnecessary waste of space. Could you guys help take a look?

ShamrockLee commented:
Just add the test cases anyway. See how OfBorg would run them.

jbedo commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -777,6 +777,36 @@ rec {
+   */
+  writeMultipleReferencesToFile = paths: runCommand "runtime-deps-multiple" {
+    # Uniquely sort input paths to reduce unnessesary packages
+    referencesFiles = map writeReferencesToFile (builtins.unique (builtins.sort paths));
```
I don't think the unique sort is necessary, `writeReferencesToFile` is deterministic so there won't be any unnecessary paths, and your dictionary below ensure uniqueness.

ShamrockLee replied:
The `lib.unique` here is trying to reduce the build time. `builtins.sort` is probably not needed.

Besides, I think that I might accidentally pushed an older version up here. `builtins.unique` should be `lib.unique`.

ShamrockLee commented:
@ofborg test references

SomeoneSerge commented:
Awesome. I'll just note that I think bash is on its way of becoming a burden here. I would really like to run some simple benchmarks because I don't have a very concrete understanding of how much of an issue this `mkLayer` is right now, and how much this change improves it, but slow as I am I wouldn't insist that you wait for my approval on merging

jbedo commented:
It's still short and understandable, so I think a shift away from bash would have more downsides than advantages.

ShamrockLee commented:
> I would also really like to run some simple benchmarks because I don't have a very concrete understanding of how much of an issue this `mkLayer` is right now, and how much this change improves it

I haven't done a serious benchmark yet. The reason why it helps is:
*   `mkLayer` produces a package by copying all the files in the output path of packages in `contents` into its own output path.
*   What `buildImage` needs is only a list of store paths of `contents` and their run-time dependencies (made with `writeReferencesToFile`), making the files `mkLayer` copies useless.
*   `writeMultipleReferencesToFile` simply dump the list of store paths to its output file, preventing the redundant copying.

When dealing with large packages like `root` (CERN ROOT, a C++ analysis framework), the use of `mkLayer` doubles the image build time and the disk space usage.

SomeoneSerge commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -761,6 +761,36 @@ rec {
+   */
+  writeMultipleReferencesToFile = paths: runCommand "runtime-deps-multiple" {
+    # Uniquely sort the input paths to reduce unnessesary rebuilds
+    referencesFiles = map writeReferencesToFile (lib.pipe (map toString paths) [ (builtins.sort builtins.lessThan) lib.unique ]);
```
Ok, so now I'm looking at https://github.com/NixOS/nixpkgs/pull/224636/files#diff-35d95fd2d33596cbcc6f411406cd7a1a847e1d86e206c7aa248732f2254b52d0R38 and I'm beginning to understand what this is meant to be used for. From what I understand, each evaluation of `writeReferencesToFile` amounts to a single `runCommand`.

Question: why should `map writeReferencesToFile (lib.pipe (map toString paths) [ (builtins.sort builtins.lessThan) lib.unique ])` ($N$ lighter runCommands) be faster than one bigger runCommand:

```nix
let
  directDeps = (lib.pipe (map toString paths) [ (builtins.sort builtins.lessThan) lib.unique ]);
  directDepsTxt = writeText "direct-deps.txt" (concatStringsSep "\n" uniqueDirectDependencies;
in writeReferencesToFile directDepsTxt;
```

ShamrockLee replied:
> Question: why should map writeReferencesToFile (lib.pipe (map toString paths) [ (builtins.sort builtins.lessThan) lib.unique ]) ($N$ lighter runCommands) be faster than one bigger runCommand:\n\nlet\n  directDeps = (lib.pipe (map toString paths) [ (builtins.sort builtins.lessThan) lib.unique ]);\n  directDepsTxt = writeText \"direct-deps.txt\" (concatStringsSep \"\\n\" uniqueDirectDependencies;\nin writeReferencesToFile directDepsTxt;

Cool! It would be perfect if we could remove the line of the direct-list packge store path .

SomeoneSerge replied:
Oh, do you mean the txt?

ShamrockLee replied:
Yes.

SomeoneSerge replied:
Nothing a little `sed` can't do 😅 I'm more concerned about having numbers on whether doing any of this is worth the effort at all. I thought about it for a moment and realized that benchmarking can be quite a bit of a job: there's nix evaluation cache, there's garbage collection...

ShamrockLee replied:
My workflow to benchmark the build of a package:

Suppose the package to be benchmarked is `hello`, a non-fixed-output derivation in Nixpkgs:

1.  Preparation
    ```sh
    # Reduce the possibility for the target package to appear in the store.
    # should we already have it in the store, the benchmark won't work.
    nix-store --gc
    cd path/to/nixpkgs
    # Get all the dependencies ready.
    nix-shell -A hello --run "sh -c :"
    ```

2.  Actual build. Add timing code around this commnd:
    ```
    nix-build --offline --no-out-link -A hello
    ```

ShamrockLee replied:
Just implemented like that.

jbedo commented:
You can use a temporary store instead of GC, and rather than a
shell to prime you can just set jobs to 0:

  TMPSTORE=$(mktemp -d)
  nix-build -A hello --store $TMPSTORE -j 0
  time nix-build -A hello --store $TMPSTORE --offline --no-out-link

ShamrockLee commented:
A temporary store would be neat!

> rather than a shell to prime you can just set jobs to 0

I don't understand this part.

The purpose of `nix-shell -A hello --run exit` is to build all the dependencies before the benchmark.

jbedo commented:
Yeah never mind, -j0 will fetch from cache but not build anything, but
the shell might actually be better if there are some dependencies that
are uncached that you don't want to include in the benchmark.

ShamrockLee commented:
Thank you for explaining. I had thought that `-j0` indicates the number of available cores just like `make`.

jbedo commented:
The equivalent of that is --cores.

I just remembered that nix-build has --check, which is perhaps the
easiest:

    ❯ hyperfine --warmup 1 "nix-build --check -A hello '<nixpkgs>'"
    Benchmark 1: nix-build --check -A hello '<nixpkgs>'
    Time (mean ± σ):      9.452 s ±  0.094 s    [User: 0.117 s, System: 0.040 s]
    Range (min … max):    9.324 s …  9.605 s    10 runs

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2978

SomeoneSerge commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -802,6 +802,28 @@ rec {
+    (writeReferencesToFile pathListFile).overrideAttrs (previousAttrs: {
+      name = "multiple-runtime-deps";
+      buildCommand = previousAttrs.buildCommand + ''
+        substituteInPlace "$out" --replace ${lib.escapeShellArg pathListFile}"
+        " ""
```
Was this linefeed at the end intentional?

ShamrockLee replied:
Yes. These is meant to remove the line (without leaving a blank line).

SomeoneSerge replied:
I see. Still looks puzzling so maybe at least a comment would be appropriate. And correctness shouldn't be a concern, because `substituteInPlace` errors out when fails to match anything in the inputs

ShamrockLee replied:
Just modified and added comments to increase readability.

SomeoneSerge reviewed (approved):
LGTM but I think that at this point we'd benefit from introducing some tests. We could expose some in singularity's passthru so that ofborg knows to rerun them as things change.

And maybe a matter for a separate PR, but a NixOS integration test executing `singularity run` with an image assembled using `singularity-tools` would be nice.

I don't know of too many singularity users in Nixpkgs so automated testing might really help with longer-term maintenance

ShamrockLee committed:
tests.trivial-builders.references: test writeClosure

Test writeClosure instead of writeReferencesToFile.

Add multiple-path test for writeReferenceClosureToFile.

Rename variables:
- references -> closures (passthru affected)
- REFERENCES -> CLOSURES

ShamrockLee commented:
> LGTM but I think that at this point we'd benefit from introducing some tests. We could expose some in singularity's passthru so that ofborg knows to rerun them as things change.

I added the test case to `tests.trivial-builders.references` long ago, but didn't push until now because `tests.trivial-builders.references` had failed to build before I added the test case.

The commit containing the test case is now pushed.

> And maybe a matter for a separate PR, but a NixOS integration test executing `singularity run` with an image assembled using `singularity-tools` would be nice.

Please take a look at #268199. Currently, I could only add image-running test for Apptainer because Sylabs SingularityCE (still) requires loop devices to mount the image to SESSIONDIR, which is not available in the Nix sandbox, no is it in `runInLinuxVM` build environment.

SomeoneSerge commented on pkgs/build-support/trivial-builders/test/invoke-writeMultipleReferencesToFile.nix:
```diff
@@ -0,0 +1,6 @@
+{ lib ? import ../../../../lib
+, pkgs ? import ../../../.. { config = { }; overlays = { }; }
+}:
```
I don't suppose we'll need these in the future? Can just do `callPackage` and refer to the test through the nixpkgs tree

SomeoneSerge replied:
H'm, I currently get this:

```
❯ nix run github:ShamrockLee/nixpkgs/write-multiple-references#tests.trivial-builders.references
...

       error: getting status of '/nix/store/n1pz69ll149p06wf51k728rrwhz8aywr-source/pkgs/build-support/trivial-builders/test/samples': No such file or directory
```

ShamrockLee replied:
> H'm, I currently get this:
> 
> ```
> ❯ nix build github:ShamrockLee/nixpkgs/write-multiple-references#tests.trivial-builders.references
> ...
> 
>        error: getting status of '/nix/store/n1pz69ll149p06wf51k728rrwhz8aywr-source/pkgs/build-support/trivial-builders/test/samples': No such file or directory
> ```

Oops! It should be `samples.nix` instead of `samples`.

SomeoneSerge replied:
AFAIU I'm getting an unexpected extra line:

```
vm-test-run-nixpkgs-trivial-builders> machine: output: --- /dev/fd/63   2023-11-29 13:38:10.831614073 +0000
vm-test-run-nixpkgs-trivial-builders> +++ /dev/fd/62    2023-11-29 13:38:10.831614073 +0000
vm-test-run-nixpkgs-trivial-builders> @@ -14,3 +14,4 @@
vm-test-run-nixpkgs-trivial-builders>  /nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27
vm-test-run-nixpkgs-trivial-builders>  /nix/store/s2f1sqfsdi4pmh23nfnrh42v17zsvi5y-libunistring-1.1
vm-test-run-nixpkgs-trivial-builders>  /nix/store/sbldylj3clbkc0aqvjjzfa6slp4zdvlj-hello-2.12.1
vm-test-run-nixpkgs-trivial-builders> +/nix/store/wxiv7cbgpncmc87qadf14nnf9ip39749-invoke-writeReferencesToFile.nix
```

ShamrockLee replied:
> AFAIU I'm getting an unexpected extra line:
> 
> ```
> vm-test-run-nixpkgs-trivial-builders> machine: output: --- /dev/fd/63   2023-11-29 13:38:10.831614073 +0000
> vm-test-run-nixpkgs-trivial-builders> +++ /dev/fd/62    2023-11-29 13:38:10.831614073 +0000
> vm-test-run-nixpkgs-trivial-builders> @@ -14,3 +14,4 @@
> vm-test-run-nixpkgs-trivial-builders>  /nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27
> vm-test-run-nixpkgs-trivial-builders>  /nix/store/s2f1sqfsdi4pmh23nfnrh42v17zsvi5y-libunistring-1.1
> vm-test-run-nixpkgs-trivial-builders>  /nix/store/sbldylj3clbkc0aqvjjzfa6slp4zdvlj-hello-2.12.1
> vm-test-run-nixpkgs-trivial-builders> +/nix/store/wxiv7cbgpncmc87qadf14nnf9ip39749-invoke-writeReferencesToFile.nix
> ```

This means that the line was missing in the file produced by `writeMultipleReferencesToFile`.

After hours of debugging, it turns out that `toString` doesn't always do string interpolation on string-interoperable stuff, especially when the input is a [path value](https://nixos.org/manual/nix/stable/language/values#type-path).

Change the `toString` to `(path: "${path}")` that constructs the `pathListFile`, and the problem is solved.

ShamrockLee replied:
> I don't suppose we'll need these in the future? Can just do `callPackage` and refer to the test through the nixpkgs tree

They are there to make it possible to `nix-build` the expression file directly, in the hope that we could run `references-test.sh` directly without having to spawn a VM to run the tests.

This currently fails, as stated in the `references-test.sh`. I think that the failure is caused by bugs yet to be discovered instead of intrinsic limitation, and I keep it in the hope that it would work in the future.

SomeoneSerge replied:
Uh-huh. So

1. The motivation for calling the file directly as opposed to through `test.trivial-builders.references` is to be able to run the `.sh` script outside the VM
2. Tests outside the VM do not work yet

RE: 1. Can't we just passthru the `invoke*` derivation in the `nixosTest`? Then we could access it by the attribute path

RE: 2. Let's just merge the part that is usable right now, and carry over the less clean bits into the later PRs?

ShamrockLee replied:
> RE: 1. Can't we just passthru the `invoke*` derivation in the `nixosTest`? Then we could access it by the attribute path

That's sounds reasonable. I implemented it in #204462, which becomes a great refactor of `tests.trivial-builders.references`. I'll try to add the tests here with minimal changes, and integrate the two PRs once any of them get merged.

ShamrockLee commented:
I removed all the unnecessarp `? import ../../../.. { ... }` from all the files changed by this PR.

SomeoneSerge reviewed (approved):
The test succeeds now. Thanks a lot @ShamrockLee!

ShamrockLee commented on pkgs/build-support/trivial-builders/test/invoke-writeReferencesToFile.nix:
```diff
@@ -1,4 +1,6 @@
-{ pkgs ? import ../../../.. { config = {}; overlays = []; } }:
+{ pkgs
+, writeReferencesToFile ? pkgs.writeReferencesToFile
+}:
 pkgs.lib.mapAttrs
-  (k: v: pkgs.writeReferencesToFile v)
+  (k: v: writeReferencesToFile v)
   (import ./sample.nix { inherit pkgs; })
```
The refactoring is provided in #204462. We don't even need the `invoke-` files to run the tests both directly and inside the build sandbox.

The diff would be large, and that's why I'm not planning to place it here.

SomeoneSerge replied:
Sure, I'd be happy with merging this as it is now

ShamrockLee commented:
Now that #204462 is merged, I refined the all-references part of `tests.trivial-builders.references`.

I also added two package tests `tests.trivial-builders.writeMultipleReferencesToFile-single` and `tests.trivial-builders.writeMultipleReferencesToFile-mixed` that doesn't require Linux VM instances to run. They compare the result of `writeMultipleReferencesToFile` with that of `writeReferencesToFile`. They are inspired by `tests.trivial-builders-writeStringReferencesToFile`.

ShamrockLee commented:
@ofborg build tests.trivial-builders.references tests.trivial-builders.writeMultipleReferencesToFile-single tests.trivial-builders.writeMultipleReferencesToFile-mixed

ShamrockLee commented:
Found two spaces in the comment. Remove one.

ShamrockLee commented:
@ofborg build tests.trivial-builders.references tests.trivial-builders.writeMultipleReferencesToFile-single tests.trivial-builders.writeMultipleReferencesToFile-mixed

ShamrockLee commented:
> Now that #204462 is merged, I refined the all-references part of tests.trivial-builders.references.\n\nI also added two package tests tests.trivial-builders.writeMultipleReferencesToFile-single and tests.trivial-builders.writeMultipleReferencesToFile-mixed that doesn't require Linux VM instances to run. They compare the result of writeMultipleReferencesToFile with that of writeReferencesToFile. They are inspired by tests.trivial-builders-writeStringReferencesToFile.

@roberth @SomeoneSerge could you help take a look at the tests?

SomeoneSerge commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -815,6 +815,36 @@ rec {
+    in
+    # Get the referenecs of pathListFile, but
+    # remove the line containing the path to pathListFile
+    (writeReferencesToFile pathListFile).overrideAttrs (previousAttrs: {
```
I've recently learned, that `exportReferencesGraph` offers a much cleaner interface when you use `__structuredAttrs = true`: https://github.com/NixOS/nixpkgs/blob/c2e4139f043eb5d1f0f679288ec965fe1c0688cf/pkgs/build-support/closure-info.nix#L16-L18 https://github.com/NixOS/nixpkgs/blob/c2e4139f043eb5d1f0f679288ec965fe1c0688cf/pkgs/build-support/closure-info.nix#L36-L38

Admittedly, this is a very surface-level comment, but I think we need to answer the following questions prior to merging:

1. Can we later rewrite `writeMultipleReferencesToFile` using the json interface for `exportReferencesGraph` while preserving the interface introduced in this PR?
2. Does `writeMultupleReferencesToFile` not duplicate e.g. `closureInfo`?

SomeoneSerge replied:
Oh, I thought this was used to compute the closure for singularity images? Sorry, I'm completely forgetting the context, this has been open for so long...

ShamrockLee replied:
> 2\. Does `writeMultupleReferencesToFile` not duplicate e.g. `closureInfo`?

Not even close.
```
nixpkgs_timeshift (9b54fb4) [$]
❯ nix build --no-link --print-out-paths --impure --expr "let pkgs = import ./. { }; in pkgs.closureInfo { rootPaths = [ pkgs.linux ]; }"
/nix/store/sd8vsqzxhq9x4hyz3fvvggss0i4wkqvi-closure-info

nixpkgs_timeshift (9b54fb4) [$]
❯ ls -A /nix/store/sd8vsqzxhq9x4hyz3fvvggss0i4wkqvi-closure-info
registration  store-paths  total-nar-size

nixpkgs_timeshift (9b54fb4) [$]
❯ cat /nix/store/sd8vsqzxhq9x4hyz3fvvggss0i4wkqvi-closure-info/registration
/nix/store/lgkpyy1b1zzh04n8b6andg8iw8vnr07m-linux-6.6.21
sha256:0m59hwzpdbv4ihbgcbgd23a8gbn3gnzx2igsygm0a6wjyz8r0pb6
131908608

0

nixpkgs_timeshift (9b54fb4) [$]
❯ cat /nix/store/sd8vsqzxhq9x4hyz3fvvggss0i4wkqvi-closure-info/store-paths
/nix/store/lgkpyy1b1zzh04n8b6andg8iw8vnr07m-linux-6.6.21

nixpkgs_timeshift (9b54fb4) [$]
❯ cat /nix/store/sd8vsqzxhq9x4hyz3fvvggss0i4wkqvi-closure-info/total-nar-size
131908608

```

ShamrockLee replied:
> Can we later rewrite `writeMultipleReferencesToFile` using the json interface for `exportReferencesGraph` while preserving the interface introduced in this PR?

I hope so. I was originally plan to implement this function with `exportReferencesGraph`. It is the confusing yet undocumented format that makes me turn to `writeReferencesToFile`.

ShamrockLee replied:
Update: I finally got an idea how to read the files generated by `exportReferencesGraph`, but the implementation will become much closer to the one based on Bash associative array (dictionary).

SomeoneSerge replied:
> Nothing is produced when passing __structuredAttrs = true;

Check `$NIX_ATTRS_JSON_FILE`/`.attrs.json`.

> with exportReferencesGraph = [ "graph" pkgs.hello ]

Looking at `closure-info.nix`, I guess this should be `exportReferencesGraph.graph = [ pkgs.hello ]` when using structured attrs

SomeoneSerge replied:
> >    2. Does writeMultupleReferencesToFile not duplicate e.g. closureInfo?
> Not even close.

(I'm half asleep and not sure what I'm looking at but) maybe the right one was `pkgs/build-support/references-by-popularity/default.nix`...

SomeoneSerge replied:
Looking at the occurrences of "`unnecessaryDrvs`" in `pkgs/build-support/docker/default.nix` I suspect they the same logic you've implemented in your `writeMultipleReferencesToFile` but scattered across `dockerTools` and `referencesByPopularity` (plus the sort thing)

SomeoneSerge replied:
I'm not sure why would `referencesByPopularity` set `exportReferences.graph` to a `Path` instead of a `List Path`, which the other derivation did. If that accepts lists, there's no reason to do the "write a file with all the root paths, then filter that file out from the closure" trick they do in `dockerTools`' `default.nix`

SomeoneSerge replied:
> Update: I finally got an idea how to read the files generated by exportReferencesGraph, but the implementation will become much closer to the one based on Bash associative array (dictionary).

Jq/python sound more promising. We don't have to rewrite now, but we do need to decide whether to introduce the new function or somehow extend the utils used by `dockerTools` (`referencesByPopularity`)

ShamrockLee replied:
> > Nothing is produced when passing __structuredAttrs = true;
> 
> Check `$NIX_ATTRS_JSON_FILE`/`.attrs.json`.
> 
> > with exportReferencesGraph = [ "graph" pkgs.hello ]
> 
> Looking at `closure-info.nix`, I guess this should be `exportReferencesGraph.graph = [ pkgs.hello ]` when using structured attrs

It's totally undocumented! Thanks for pointing out.

ShamrockLee replied:
@SomeoneSerge Thank you for telling me about the beautiful `exportReferencesGraph` interface with `__structuredAttrs` enabled. Now the implementation is as elegant as a Nix expression!

roberth commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -815,6 +815,36 @@ rec {
         writeDirectReferencesToFile (writeText "string-file" string);
 
 
+  /**
+   * Write the references (i.e. the runtime dependencies in the Nix store) of the elements in `paths' to a file.
+   */
```
Could you move this to the manual, `doc/build-helpers/trivial-build-helpers.chapter.md`?

What to do was unclear, so I've opened
- https://github.com/NixOS/nixpkgs/pull/294944

SomeoneSerge replied:
I just checked and `referencesByCount` expects `rootPaths` to be string-like

roberth reviewed (changes requested):
The name suggests that it only writes the references, but it writes the whole closure.
While this is a pre-existing problem with `writeReferencesToFile`, we should really use a better name.
(It surely caught me off guard: https://github.com/NixOS/nixpkgs/pull/178717/files#r1519520762)

Suggestion: `writeReferenceClosureToFile`

We could then also start deprecating `writeReferences`, which caused all this, but that doesn't have to be done in this PR.

ShamrockLee commented:
> Suggestion: `writeReferenceClosureToFile`

How could we distinguish between this function and the current `writeReferencesToFile`?

ShamrockLee committed:
writeClosure: init, replacing writeReferencesToFile

Replace writeReferencesToFile with writeClosure.

Make writeClosure accept a list of paths instead of a path.

Re-implement with JSON-based exportReferencesGraph interface provided by
__structuredAttrs = true.

Reword the documentation.

Co-authored-by: Valentin Gagarin <valentin.gagarin@tweag.io>
Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>
Co-authored-by: Someone Serge <sergei.kozlukov@aalto.fi>

ShamrockLee committed:
nixosTests.nixops: remove dead code from Nix expression

ShamrockLee committed:
referenceByPopularity: rename in comment writeReferencesToFile -> writeClosure

ShamrockLee committed:
singularity-tools.buildImages: use writeClosure

ShamrockLee committed:
dockerTools.buildImage: writeReferencesToFile -> writeClosure

ShamrockLee committed:
ociTools.buildContainer: writeReferencesToFile -> writeClosure

ShamrockLee commented:
> The name suggests that it only writes the references, but it writes the whole closure. While this is a pre-existing problem with `writeReferencesToFile`, we should really use a better name. (It surely caught me off guard: https://github.com/NixOS/nixpkgs/pull/178717/files#r1519520762)
> 
> Suggestion: `writeReferenceClosureToFile`
> 

As there are also `writeDirectReferencesToFile`, should we call it `writeReferenceClosureToFile` or `writeClosureReferencesToFile`?

> We could then also start deprecating `writeReferences`, which caused all this, but that doesn't have to be done in this PR.

How about deprecating `writeReferencesToFile` in favor of `writeReferenceClosureToFile`, and re-implement `writeDirectReferencesToFile` to accept a list of paths?

SomeoneSerge commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -625,18 +625,28 @@ rec {
-        read nrRefs
-        for ((i = 0; i < nrRefs; i++)); do read ref; done
-      done < graph
+      nix --extra-experimental-features "nix-command" \
+        --offline \
+        eval --eval-store "$PWD" --store "$PWD" --impure --expr ${lib.escapeShellArg ''
+          let
+            lib = import ${nixpkgs}/lib;
+            rawAttrs = lib.importJSON ./.attrs.json;
+            references = lib.sort lib.lessThan (map (lib.getAttr "path") rawAttrs.graph);
+          in
+          lib.concatStringsSep "\n" references
+        ''} \
+        --raw > "$out"
```
I think it's better not to introduce a dependency on nix... As mentioned above, `closure-info` contains examples of using `jq` and `referencesByPopularity` - of using python's `json` - to read the output of `exportReferencesGraph`

ShamrockLee replied:
Alright.

ShamrockLee replied:
Addressed by parsing and sorting with jq.

roberth reviewed (approved):
Typo, but lgtm!

ShamrockLee commented:
BTW, should we also make `writeDirectReferencesToFile` accept a list of paths?

Implementation-wise, it's natural for `writeReferenceClosureToFile` to support a list of paths, as the structural `exportReferencesGraph` itself already accepts path list, and the result don't have to be de-duplicated. However, the *direct references* are provided for each closure paths specifically, which means that we'll have to mix and de-duplicate manually if we would like `writeDirectReferencesToFile` to support a list of paths.

ShamrockLee commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -581,7 +581,7 @@ rec {
 
   # Docs in doc/builders/special/makesetuphook.section.md
-  # See https://nixos.org/manual/nixpkgs/unstable/#sec-pkgs.makeSetupHook
+  # See https://nixos.org/manual/nixpkgs/unstanixble/#sec-pkgs.makeSetupHook
```
It was inserted accidentally by some mouse clicks on VSCode.

ShamrockLee commented:
Add a release note, and a warning/throwing message for `writeReferencesToFile`.

SomeoneSerge commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -625,18 +625,21 @@ rec {
-        read nrRefs
-        for ((i = 0; i < nrRefs; i++)); do read ref; done
-      done < graph
+      jq -r ".graph | sort_by(.path) | .[].path" ./.attrs.json > "$out"
```
Yaaay this looks so simple now!

Nit: looking at https://nixos.org/manual/nix/unstable/language/advanced-attributes.html?highlight=__structured#adv-attr-structuredAttrs, I guess the "public" door is `$NIX_ATTRS_JSON_FILE` rather than the path itself.

SomeoneSerge reviewed (approved):
Aside from the comment above, I think this is ready and this might prove a good primitive.

That said, for the further steps... I'm wondering if we should ask `@grahamc` (introduced `streamLayeredImage`) or `@utdemir` (authors the `unnecessaryDrvs` line in git-blame) for a comment on `writeReferenceClosureToFile` versus `referencesByPopularity`?

- The latter could be extended to accept the same interface.
- I didn't read the code but I guess they're both dominated by the sort time.
- `writeReferenceClosureToFile` sorts by hashes and guarantees deterministic outputs but is very sensitive to perturbations.
- `referencesByPopularity` sorts by more persistent features so it should be less sensitive?
- `referencesByPopularity` and `dockerTools.streamLayeredImage` could be refactored to use lists of roots

ShamrockLee commented:
> `writeReferenceClosureToFile` sorts by hashes and guarantees deterministic outputs but is very sensitive to perturbations.

This reminds me that we could sort the paths before passing to `exportReferencesGraph` to avoid unnecessary rebuilds caused by perturbation changes.

Aside from that, I also filter out `null` and empty string before passing to `exportReferencesGraph` to make `writeReferenceClosureToFile` more resilient. `null` is often passed into the `override` interface (`callPackage` or `<pkg>.override`) to "cancel out" package injection, which would cause current (on-`master`) implementation of `writeReferencesToFile` and `writeDirectReferencesToFile` to fail. This use case is properly handled in the proposed `writeReferenceClosureToFile` implementation.

SomeoneSerge commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -625,18 +625,24 @@ rec {
     {
-      exportReferencesGraph = [ "graph" path ];
+      __structuredAttrs = true; # Get the cleaner exportReferencesGraph interface
+      exportReferencesGraph.graph =
+        lib.sort lib.lessThan (
+          map (p: "${p}") (
+            lib.filter (p: p != null && p != "") paths));
```
1. I imagine this means that in order to _evaluate_ anything about `exportReferencesGraph.graph` (e.g. `(head exportReferencesGraph.graph).pname)`, or length, or whatever) we first have to run the _builds_ for all of the `paths`. Idk if that fundamentally changes anything.
2. Can be rewritten as a flatter expression using e.g. `lib.pipe`

<details>

```
lib.pipe paths [
  (builtins.filter (p: p != null && p != "")
  (map (p: "${p}")) # or builtins.toString?
  (lib.sort lib.lessThan)
]
```

</details>

ShamrockLee replied:
> 1. I imagine this means that in order to _evaluate_ anything about `exportReferencesGraph.graph` (e.g. `(head exportReferencesGraph.graph).pname)`, or length, or whatever) we first have to run the _builds_ for all of the `paths`. Idk if that fundamentally changes anything.

These string-like stuff will get interpolated during derivation evaluation, even if I don't interpolate them manually. They are the *dependencies* of the result derivation, and will be realized before building the result derivation. That's essentially how Nix works.

> 2. Can be rewritten as a flatter expression using e.g. `lib.pipe`
> 
> <details>
> 
> ```
> lib.pipe paths [
>   (builtins.filter (p: p != null && p != "")
>   (map (p: "${p}")) # or builtins.toString?
>   (lib.sort lib.lessThan)
> ]
> ```
> 
> </details>

We cannot use `toString` instead of string interpolation in this context. `toString ./.` simply evaluates to the absolute path to that source directory instead of a store path. This case is covered under `tests.trivial-builders.references`, as one of the sample attribute `tests.trivial-builders.references.samples.path` is a Nix-Language path. I made similar mistake when working on #204462; it took me hours to debug.

As for `lib.pipe`, I like the style pretty much, but I'm not sure if other people would think of it as an unnecessary evaluation overhead or not.

ShamrockLee replied:
Alright, maintainability matters. Let's embrace a flatter stream-processing expression.

ShamrockLee commented:
Caught a dangling `writeReferencesToFile`. Fixed.

ShamrockLee commented on pkgs/build-support/oci-tools/default.nix:
```diff
@@ -72,7 +72,7 @@
       mkdir -p $out/rootfs/{dev,proc,sys}
       cp ${config} $out/config.json
-      xargs tar c < ${writeReferencesToFile args} | tar -xC $out/rootfs/
+      xargs tar c < ${writeReferenceClosureToFile args} | tar -xC $out/rootfs/
```
@spacekookie How did you made it to feed `writeReferencesToFile` a list? Is `args` meant to be a list or a path?

My best guess is that the singleton were flattened into the list assigned to `exportReferencesGraph`, causing the result to be identical to that with a single path passed.

SomeoneSerge commented on nixos/tests/nixops/default.nix:
Why are we removing this?

ShamrockLee replied:
It's a piece of dead code inside a let-in block. Removing it doesn't change the expression.

The removal is part of the transition from `writeReferencesToFile` to the proposed builder.

SomeoneSerge commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -625,18 +625,29 @@ rec {
-        read nrRefs
-        for ((i = 0; i < nrRefs; i++)); do read ref; done
-      done < graph
+      jq -r ".graph | sort_by(.path) | .[].path" "$NIX_ATTRS_JSON_FILE" > "$out"
```
Shaving the yak: `jq -r ".graph | map(.path) | sort_by(.) | .[]"`

ShamrockLee replied:
Now it's `jq -r ".graph | map(.path) | sort | .[]"`

SomeoneSerge commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -625,18 +625,29 @@ rec {
-      exportReferencesGraph = [ "graph" path ];
+      # Get the cleaner exportReferencesGraph interface
+      __structuredAttrs = true;
+      exportReferencesGraph.graph = lib.pipe paths [
+        # Filter out common exceptions
+        (lib.filter (p: p != null && p != ""))
+        # String interpolation
+        (map (p: "${p}"))
+        # Sort paths to avoid unnecessary rebuild
+        (lib.sort lib.lessThan)
+      ];
```
How about... we move `lib.sort` into a follow-up PR and merge this one in its simplest form?

ShamrockLee replied:
We could move the whole list processing part to the following PR should it requires further discussion.

SomeoneSerge commented on doc/build-helpers/trivial-build-helpers.chapter.md:
```diff
@@ -557,14 +557,18 @@ This creates a derivation with a directory structure like the following:
+
+## `writeReferenceClosureToFile` {#trivial-builder-writeReferenceClosureToFile}
+
+Writes the closure of transitive dependencies of the given paths to a file.
```
Nit: to me this almost reads like we're given "paths" which are "paths to a file", rather than "we write to a file".

Just so I propose at least something: "given a list of store paths (???), writes the closure of their transitive dependencies to a file".

ShamrockLee replied:
It's directly taken from the documentation of `writeReferencesToFile`, with "the given path" -> "the given paths".

The new version sounds more clearly.

ShamrockLee replied:
I modify the doc. Please take a look.

SomeoneSerge commented on doc/build-helpers/trivial-build-helpers.chapter.md:
```diff
@@ -557,14 +557,18 @@ This creates a derivation with a directory structure like the following:
-Writes the closure of transitive dependencies to a file.
+Deprecated. Use [`writeReferenceClosureToFile`](#trivial-builder-writeReferenceClosureToFile) instead.
+
+## `writeReferenceClosureToFile` {#trivial-builder-writeReferenceClosureToFile}
```
Also, `writeReferencesClosureToFile` is a bit of a mouthful, I can imagine myself typoing or forgetting if references are plural

ShamrockLee commented:
Here are some candidates to name the new build helper.

[`"Closure"`](https://nixos.org/manual/nix/stable/glossary#gloss-closure) is defined in Nix Manual.

> The closure of a store path is the set of store paths that are directly or indirectly “reachable” from that store path; that is, it’s the closure of the path under the references relation. For a package, the closure of its derivation is equivalent to the build-time dependencies, while the closure of its output path is equivalent to its runtime dependencies. For correct deployment it is necessary to deploy whole closures, since otherwise at runtime files could be missing. The command nix-store --query --requisites prints out closures of store paths.
> 
> As an example, if the [store object](https://nixos.org/manual/nix/stable/glossary#gloss-store-object) at path P contains a [reference](https://nixos.org/manual/nix/stable/glossary#gloss-reference) to a store object at path Q, then Q is in the closure of P. Further, if Q references R then R is also in the closure of P.

"Requisties" from `nix-store --query --requisties` could be a similar choice, but "indirect references" should be ruled out due to ambiguity.

As for the suffix, do we really need to emphasize `ToFile`?

Candidate list:

- `writeReferenceClosure`
    - Inconsistent with `writeDirectReferencesToFile`, and users have to memorize when to use the singular / plural form of the word "reference".
- `writeClosureReferences`
    - Consistent with `writeDirectReferencesToFile`.
    - Is it dealing with closure or references?
- `writeClosure`
    - The shortest among all.
    - I'm not sure if there would be confusion with the NAR dump.
- `writeRequisties`
    - An alternative to `writeClosure`, indicating the similarity in behavior with `nix-store --query --requisties`.
    - Do we want the term "closure" and "requisties" to coexist?

SomeoneSerge commented:
> `writeClosureReferences`

Well at least with this one I'm confident I can remember all the letters

SomeoneSerge commented on doc/build-helpers/trivial-build-helpers.chapter.md:
```diff
@@ -557,14 +557,18 @@ This creates a derivation with a directory structure like the following:
-This produces the equivalent of `nix-store -q --requisites`.
+## `writeClosure` {#trivial-builder-writeClosure}
+
+Given a list of [store paths](https://nixos.org/manual/nix/stable/glossary#gloss-store-path) (or string-like expressions coercible to store paths), it writes the combined [closure](https://nixos.org/manual/nix/stable/glossary#gloss-closure) of them to a text file.
```
```suggestion
Builds a line-separated file with all of the [store paths](https://nixos.org/manual/nix/stable/glossary#gloss-store-path) in the [closure](https://nixos.org/manual/nix/stable/glossary#gloss-closure) of a list of root paths
```

SomeoneSerge replied:
Could also make it


```suggestion
Builds a line-separated file with [references](https://nixos.org/manual/nix/stable/glossary#gloss-reference) to all of the [store paths](https://nixos.org/manual/nix/stable/glossary#gloss-store-path) in the [closure](https://nixos.org/manual/nix/stable/glossary#gloss-closure) of a list of root paths
```


Idk

SomeoneSerge replied:
"root paths" is kind of informal though

SomeoneSerge replied:
@fricklerhandwerk sorry to ping but we're discussing names and documentation, would you like to share an opinion?

ShamrockLee replied:
Thank you for helping!

"References of all store paths in the closure" is equivalent to "the closure". I think it might be confusing to mention *reference* here.

fricklerhandwerk replied:
```suggestion
Given a list of [store paths](https://nixos.org/manual/nix/stable/glossary#gloss-store-path) (or string-like expressions coercible to store paths), write their  collective [closure](https://nixos.org/manual/nix/stable/glossary#gloss-closure) to a text file.
```

I (strongly) recommend using imperative in reference documentation, primarily for brevity. "Do this, do that."

@SomeoneSerge the original structure was actually better because the prose matches the functional interface: given <input> make <output>.

@ShamrockLee [to combine](https://en.wiktionary.org/wiki/combine#Etymology) means "to put together two things". Since the input is a list, the word should be along the lines of [to compose](https://en.wiktionary.org/wiki/compose#Etymology) or [to collect](https://en.wiktionary.org/wiki/collect#Etymology_1).

SomeoneSerge replied:
Yes, "given x, do y" is actually my preference too; what I intended with the suggestion is to specify the format ("line-separated") and get rid of the "combined closure".

On the latter note, I'd just drop "combined": "the closure of a set of paths" is already meaningful, there's no need for extra adjectives. The extra adjectives can also be harmful if they are perceived as technical terms which they are not

SomeoneSerge reviewed (approved):
I think you chose the more extreme name. Personally, I like this better, but I'll wait a bit for someone else to defend "References". I suppose "References" is a more explicit reference (jaja) to the the GC and dependency detection: you "writeReferencesToFile" to retain a runtime dependency on something you've seen at build time

roberth commented:
In Nix, we have a default relation we refer to when talking about "the" closure, which is the relation of references, so I think the name `writeClosure` is fine.

> I suppose "References" is a more explicit reference (jaja) to the the GC and dependency detection: you "writeReferencesToFile" to retain a runtime dependency on something you've seen at build time

For this use case, taking the direct references is sufficient. The GC knows that the whole closure is reachable from that starting set.
"Direct" is implied when talking about [references](https://nixos.org/manual/nix/stable/glossary.html?highlight=reference#gloss-reference), but I've had to use that term when I added `writeDirectReferencesToFile`, to make it distinct from the somewhat badly named `writeReferencesToFile`.

ShamrockLee commented:
Just fixed a commit message. Sorry for another force push.

SomeoneSerge commented on pkgs/build-support/trivial-builders/test/writeClosure-union.nix:
(I've been waiting for the darwin check and couldn't resist posting a) Nit: might as well use `nixfmt-rfc-style` for the new files:) But let's maybe not reset Ofborg

ShamrockLee replied:
That RFC style guide is a long one. Since it doesn't affect evaluation,we could reformat in subsequent PRs.

ShamrockLee replied:
BTW, `tests.trivial-builders.references` currently evaluates to `{ }` on non-Linux to make OfBorg build always green. We don't have to wait for that.

SomeoneSerge commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -625,18 +625,22 @@ rec {
-        read nrRefs
-        for ((i = 0; i < nrRefs; i++)); do read ref; done
-      done < graph
+      jq -r ".graph | map(.path) | sort | .[]" "$NIX_ATTRS_JSON_FILE" > "$out"
```
Reminder: you can now open a PR with the `lib.sort` feature you proposed earlier. I still don't understand the implications this would have

ShamrockLee replied:
#300722 it is!

ShamrockLee commented:
> Does `writeMultupleReferencesToFile` not duplicate e.g. `closureInfo`?

@SomeoneSerge, sorry that I made it wrong. `"${pkgs.writeClosure paths}"` *is* a duplication of `"${pkgs.closureInfo paths}"`.

I didn't realize that the whole closure of `pkgs.linux` is only itself.

SomeoneSerge commented:
> "${pkgs.writeClosure paths}" is a duplication of "${pkgs.closureInfo paths}".

I thought we concluded someplace else that `writeClosure` had a different signature?

ShamrockLee commented:
> I thought we concluded somewhere else that `writeClosure` had a different signature

They have different signatures. `pkgs.closureInfo` is not a drop-in replacement of `pkgs.writeClosure`. As for their functions, one can use `pkgs.closureInfo` to achieve a functionality similar to that of `pkgs.writeClosure`.

I said they are "not even close" (https://github.com/NixOS/nixpkgs/pull/178717#discussion_r1525164592) based on the content of the file `"${pkgs.closureInfo { storePaths = [ pkgs.linux ]; }/store-paths"`. It contains only one line

```
nixpkgs_timeshift (9b54fb4) [$]
❯ cat /nix/store/sd8vsqzxhq9x4hyz3fvvggss0i4wkqvi-closure-info/store-paths
/nix/store/lgkpyy1b1zzh04n8b6andg8iw8vnr07m-linux-6.6.21
```

I also thought the closure of `pkgs.linux` would be much bigger than that, but now I know I was wrong.
